ldap/servers/slapd/back-ldbm/db-mdb/mdb_import.c

```diff
@@ -621,6 +621,14 @@ dbmdb_import_monitor_threads(ImportJob *job, int *status)
             sprintf(p, "recent rate %.1f/sec, ",
                          job->recent_progress_rate);
             import_log_notice(job, SLAPI_LOG_INFO, "dbmdb_import_monitor_threads", "%s", buffer);
+            for (current_worker = job->worker_list; current_worker != NULL; current_worker = current_worker->next) {
```

Comment:
Each of these workers are seperate threads updating their status yes? There are no locks here, and certainly no atomics. So all of these workers will report incorrect/conflicting information. I'm concerned about this, because this will further add issues.

Reply:
Although you are right, the logged statistics may sometime not be exact, I am not  concerned about it:
As we are aggregating data on a quite long period of time (about 3 hours on a 10M entries import),
in most case (i.e last change missed or mismatch between second and nanoseconds) the error will not be perceptible ( a few seconds against a few thousands of seconds ) and in all causes it will not change the general trend.

Furthermore that is already the case with the way both bdb and lmdb monitoring threads
 computes the import progress statistics (no lock/atomic operation are used). 
I am much more concerned about adding lock ( or even atomic operation)  in the import writer thread that is already the one that is limiting the performance. 

That said, I will add a comment around the progress rate and mdb_stat_summarize log_notice telling that the statistics may be a bit off because of the lack of lock (so that we do not waste time if someone complains)

Reply:
I'm not going to accept this as an answer. Just because the existing code does this, does not make it "correct". It means we have an opportunity to fix up past mistakes. 

Additionally, we are now seeing non-ordered arches like aarch64 become more prevalent and this code will have many more issues on those architectures.

Either user atomics to set and read the values, use a mutex with a try_lock on the reporter to prevent blocking the importer, or don't have stats at all.

This code would not pass tsan, it would probably have issues with asan if you push it. I don't think it meets the high standards we should have as a team for a high reliability system.

Reply:
I am afraid that I disagree with your arguments:
   - "High quality standard" and "correct code" means adapting the solution to the needs and avoiding using a sledgehammer to break a nuts.
   - The non-ordered arches argument is irrelevant: Whatever the order is, a read will either returns the
     right value or the old one (and old is not so old compared to 1 second unless the hardware architecture is seriously broken).
     And we just do not care whether any of these value is the current or the old one.
   - ASAN will not report any issue as the memory is allocated and initialized.
   - IMHO We have a lot of work to do before getting a clean tsan report (a bit like the cleanup I have recently done 
      about covscan report) but it is easy to handle the assumed races (by adding a suppress comment) 
      (Anyway I think it is work we will have to do when working of getting a clean tsan status)

And none of the solutions you are proposing looks really appealing:
  - try_lock will not help. The issue is not about blocking (the monitoring thread was only trying to read the data every 20 seconds so 
    there will almost never be any lock contention) but we are speaking about locking or unlocking 300000 time per seconds
    and adding all these fences have an impact.
  - I do not like either the idea about not providing theses statistics:
    I felt the need to add these statistics to try to understand why in my test with 10M entries
    lmdb import is 3 time slower than bdb one (and the statistics helped to determine that ldbm db was the bottleneck)
    People will complain that the import is not fast enough (It always happens, it is just a recurring fact.)
    and I think we should have a way to determine whether the issue is the write performance or because for
    some reason data does not get provided fast enough.
    so IMHO it is better to have the info out of the box than working with the pressure of a customer.

That said as a conciliation effort:
 - I moved the calloc out of the writer thread and do it when initializing the thread info struct (i.e before creating the threads)
 - statistics are now logged (and read from the thread info struct) only once:
   after the threads have finished their job (so the data are no more moving).

I hope these two changes does help to ease your concerns

Reply:
I'm not going to click approve on multi threaded code, that doesn't use basic thread synchronisation primitives. It's as simple as that. There are plenty of ways to make this correct, simple and fast. 

If you aim for fast above all else, you will forever sacrifice correctness and simplicity. Just because you *believe* there are no issues here today, does not make that true. There are multi-threaded code with atomics and locks that do 100,000's operations per second without problem. It can be done. 

Either add atomics, or use a lock with try_lock to prevent blocking on the workers/reader. 

I will not approve code that does not adhere to basic memory and thread safety principles.

Reply:
This just does not make sense.
There is nothing to synchronize as reader thread now waits than the writer thread is dead before reading the data !!!

Reply:
@progier389 , diving into that very technical discussion, I have a dummy question. My understanding is that the workers manage their own statistics. Then when all of them, have completed their jobs and are dead, another thread (provider, writer.. ?) reads their statistics, do a summary of them then log them. Am I correct ?

Reply:
Exactly: 
Each worker_list threads (including the provider thread, the worker threads and the writer thread) update its own set of statistics.
and the main thread (i.e the import main thread) loops 
   to compute and log the import progression rate (that is the legacy code that I have not changed)
until all threads are completed 
then read the new statistic,  compute the ratio between the different statistics values then log them

Reply:
Thanks @progier389 for your explanations, so it was a bit more complex than I initially thought.

My understanding is that there are two logs:  import statistics, import rate. 
I agree with @Firstyear that without sync between writers (workers/provider/writer) and reader (main import) there is a risk of invalid import statistics and potential others nasty things. I have no strong opinion between the two options: locking (with a risk of perf impact) and final logging without locking (only one status). In case final logging appear to be unsuffisient we could open a new ticket to implement intermediate logging, at the moment it looks an easy way to address both concerns: exact stat and no perf impact.

Regarding import rate, it is not modified by the PR. @firstyear, is it this part that you would like to improve ?

Reply:
The problem here is that you can not have multi-threaded coded, synchronising data without using atomics or mutexes. You are risking introducing undefined behaviour, race conditions, crashes. 

I know you are a really exceptional engineer @progier389 but that doesn't mean we can flaunt the rules. When we work on a project like 389-ds we have to bring our very best to ensure the best results for our users and that means following "correct, simple, fast". Today the lack of synchronisation primitives violates "correct". 

Please just add mutexes here. You can use try-lock to avoid blocking. Performance here isn't a problem as already proven by my use of try-lock in the connection table previously.

Reply:
Even if I still disagree with your arguments, I realized that since I moved the log of the statistics
 at the end of the import, there is no more any good reason to log them in the monitoring thread.
So I moved the log into the writing thread and now the statistics counters are pure local variables
 and we can stop that fruitless argument that is only delaying the fix of a reproducible crash !

Please, can you review the last changes ?